proposals/0000-no-toplevel-field-selectors.rst

```diff
@@ -0,0 +1,168 @@
+declarations to generate toplevel field accessor functions for records, such
+that the user can define their own via Generic or positional extraction.
+
+Motivation
+------------
+
+There have been proposals to extend the usage of records in Haskell
+(DuplicateRecordFields, OverloadedRecordFields, DeriveGeneric to name a few),
+but none of them address the original issue where record fields steal the
+toplevel function name for each field for selector purposes. This proposal
+intends to open the design space to this issue by removing the generation of
+these toplevel selector functions.
+
+Possible Alternative Use Cases for Record Names
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+After removing the toplevel selector names, the field names could be used as
+bindings for other values.
+
+- A `generic-lens` equivalent for toplevel lenses
+- namespaced accessors
+- namespaced lenses
+- overloaded-labels based accessors / lenses
+- row types
+- https://github.com/ghc-proposals/ghc-proposals/pull/158
+- ...
+
+Proposed Change Specification
+-----------------------------
+
+Record definition no longer defines toplevel selector functions when the
+``NoToplevelFieldSelectors`` extension is enabled. A
+``NoToplevelFieldSelectors`` pragma for records will also be available, such
+that this extension can be enabled on a per-record basis.
+
+Record construction/update syntax and pattern matching will work as before, the
+disambiguation handled by ``DuplicateRecordFields``. The necessary selectors
+will be part of a ``Record(..)`` export, or can also be named one by one. They
+always have to be associated with the data type though, because there is no more
+toplevel selector.
+
+A function for Template Haskell to go from a record field name to a selector for
+that field will be provided, because it's not possible anymore to go from field
+name directly to selector.
+
+A new `TH` function is added which takes two `Name` and returns an `Exp` which
+is equivalent to `a -> b`, where `a` is the record type specified by the second
+argument, and `b` the contents of the field specified by the first argument.
+
+Example
+^^^^^^^
+
+Given a data structure
+
+    data Foo = Foo { bar :: Int, baz :: String }
+
+The following will be available:
+
+- the type ``Foo``
+- the constructor ``Foo``
+- the two functions ``bar`` and ``baz``
+- the names ``bar`` and ``baz`` for record construction (``Foo { bar = 3, baz = "foo" }``)
+- the names ``bar`` and ``baz`` for ``RecordWildCards``
+
+If the language extension ``NoToplevelFieldSelectors`` is enabled for the module
+or ``Foo`` specifically, all of the above will be generated, except for the two
+functions ``bar`` and ``baz``.
+
+The following declarations would be equivalent:
+
+::
+
+    {-# LANGUAGE NoToplevelFieldSelectors #-}
+
+    data Foo = Foo { bar :: Int, baz :: String }
+
+::
+
+    {-# NoToplevelFieldSelectors #-}
+    data Foo = Foo { bar :: Int, baz :: String }
```

Comment:
You suggest `{-# NoToplevelFieldSelectors Foo #-}` instead?

Reply:
That would be consistent with INLINE. Otherwise to be consistent with UNPACK You'd put it after the `data` (or after the `=` or somewhere generally around there), and not have a named variant. It's really the existence of the named and unnamed variant that's a bit weird. (That said, as below, I'd ditch both.)